python/lsst/scarlet/lite/blend.py

```diff
@@ -0,0 +1,276 @@
+            bands=self.observation.bands,
+            yx0=cast(tuple[int, int], self.observation.bbox.origin[-2:]),
+        )
+
+        if use_flux:
+            for src in self.sources:
+                if src.flux is None:
+                    raise ValueError(
+                        "Some sources do not have 'flux' attribute set. Run measure.conserve_flux"
+                    )
+                src.flux.insert_into(model)
+        else:
+            for component in self.components:
+                component.get_model().insert_into(model)
+            if convolve:
+                return self.observation.convolve(model)
+        return model
+
+    def _grad_log_likelihood(self) -> Image:
+        """Gradient of the likelihood wrt the unconvolved model"""
+        model = self.get_model(convolve=True)
+        # Update the loss
+        self.loss.append(self.observation.log_likelihood(model))
+        # Calculate the gradient wrt the model d(logL)/d(model)
+        result = self.observation.weights * (model - self.observation.images)
+        result = self.observation.convolve(result, grad=True)
+        return result
+
+    @property
+    def log_likelihood(self) -> float:
+        """The current log-likelihood"""
+        if len(self.loss) == 0:
+            return self.observation.log_likelihood(self.get_model(convolve=True))
+        return self.loss[-1]
+
+    def fit_spectra(self, clip: bool = False) -> Blend:
+        """Fit all of the spectra given their current morphologies
+
+        Parameters
+        ----------
+        clip:
+            Whether or not to clip components that were not
+            assigned any flux during the fit.
+
+        Returns
+        -------
+        blend:
+            The blend with updated components is returned.
+        """
+        from .initialization import multifit_spectra
+
+        morphs = []
+        spectra = []
+        factorized_indices = []
+        model = Image.from_box(
+            self.observation.bbox,
+            bands=self.observation.bands,
+            dtype=self.observation.dtype,
+        )
+        for idx, component in enumerate(self.components):
+            if hasattr(component, "morph") and hasattr(component, "spectrum"):
+                component = cast(FactorizedComponent, component)
+                morphs.append(component.morph)
+                spectra.append(component.spectrum)
+                factorized_indices.append(idx)
+            else:
+                model.insert(component.get_model())
+        model = self.observation.convolve(model, mode="real")
+
+        boxes = [c.bbox for c in self.components]
+        fit_spectra = multifit_spectra(
+            self.observation,
+            [Image(morph, yx0=cast(tuple[int, int], bbox.origin)) for morph, bbox in zip(morphs, boxes)],
+            model,
+        )
+        for idx in range(len(morphs)):
+            component = cast(FactorizedComponent, self.components[factorized_indices[idx]])
```

Comment:
I'm not sure what you mean here, but it may be a misunderstanding about what this function does. This is _not_ the function that fits the spectra during optimization, it is only used to generate an initial set of spectra that are fit jointly given the initial morphologies.